### PR TITLE
MODBULKOPS-379 - TD: Graceful handling S3 errors

### PR DESCRIPTION
**Pull request by obozhko-folio**
[MODBULKOPS-308](https://folio-org.atlassian.net/browse/MODBULKOPS-308) - TD: Graceful handling S3 errors

## Purpose
Errors related to the S3 override other errors even if they are not related to S3. The error handling logic needs to be modified so that S3-related errors do not hide other types of errors.

## Approach
Add catch for S3ClientException.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 1]
```diff
@@ -174,9 +175,13 @@ public BulkOperation uploadCsvFile(EntityType entityType, IdentifierType identif
           operation.setProcessedNumOfRecords(numOfLines);
           operation.setMatchedNumOfRecords(numOfLines);
 
-        } catch (Exception e) {
-          log.error(ERROR_STARTING_BULK_OPERATION + e.getCause());
+        } catch (S3ClientException e) {
           errorMessage = ERROR_NOT_UPLOAD_FILE_S3_INVALID_CONFIGURATION;
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:179]: Taking into account this error message will be available in dev console and used for troubleshooting, it makes sense to add information about exception cause here.

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 2]
```diff
@@ -191,9 +196,13 @@ public BulkOperation uploadCsvFile(EntityType entityType, IdentifierType identif
       try {
         var linkToTriggeringFile = remoteFileSystemClient.put(multipartFile.getInputStream(), operation.getId() + "/" + multipartFile.getOriginalFilename());
         operation.setLinkToTriggeringCsvFile(linkToTriggeringFile);
-      } catch (Exception e) {
-        log.error(ERROR_STARTING_BULK_OPERATION + e);
+      } catch (S3ClientException e) {
         errorMessage = ERROR_NOT_UPLOAD_FILE_S3_INVALID_CONFIGURATION;
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:200]: Еру ыфьу фы https://github.com/folio-org/mod-bulk-operations/pull/316/files#r1855540575

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 3]
```diff
@@ -191,9 +196,13 @@ public BulkOperation uploadCsvFile(EntityType entityType, IdentifierType identif
...
+      } catch (Exception e) {
+        errorMessage = format(FILE_UPLOADING_FAILED_REASON, e.getMessage());
+      }
+      if (nonNull(errorMessage)) {
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:204]: Handling of error message outside catch-block makes code dirty. Please, make error messages logging for particular exception inside corresponding catch block.

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 4]
```diff
@@ -301,26 +312,35 @@ public void confirm(BulkOperation operation)  {
...
-      dataProcessing.setProcessedNumOfRecords(processedNumOfRecords);
-      dataProcessingRepository.save(dataProcessing);
+      updateProcessed(dataProcessing, operation, processedNumOfRecords);
 
-      operation.setApproach(IN_APP);
-      operation.setStatus(OperationStatusType.REVIEW_CHANGES);
-      operation.setProcessedNumOfRecords(processedNumOfRecords);
       bulkOperationRepository.findById(operation.getId()).ifPresent(op -> operation.setCommittedNumOfErrors(op.getCommittedNumOfErrors()));
+    } catch (S3ClientException e) {
+      errorMsg = ERROR_NOT_CONFIRM_CHANGES_S3_ISSUE;
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:319]: The same as https://github.com/folio-org/mod-bulk-operations/pull/316/files#r1855540575

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 5]
```diff
@@ -301,26 +312,35 @@ public void confirm(BulkOperation operation)  {
...
-      dataProcessingRepository.save(dataProcessing
-        .withStatus(StatusType.FAILED)
-        .withEndTime(LocalDateTime.now()));
-      operation.setStatus(OperationStatusType.FAILED);
-      operation.setEndTime(LocalDateTime.now());
-      operation.setErrorMessage(ERROR_NOT_CONFIRM_CHANGES_S3_ISSUE);
+      errorMsg = e.getMessage();
     } finally {
+      if (nonNull(errorMsg)) {
+        dataProcessingRepository.save(dataProcessing
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:324]: Using error message as flag makes code dirty. I would propose extract this logic in separate method and use it with in each catch-block with different error messages.

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 6]
```diff
@@ -357,21 +378,24 @@ private void confirmForInstanceMarc(BulkOperation operation)  {
...
-        dataProcessing.setProcessedNumOfRecords(processedNumOfRecords);
-        dataProcessingRepository.save(dataProcessing);
 
-        operation.setApproach(IN_APP);
-        operation.setStatus(OperationStatusType.REVIEW_CHANGES);
-        operation.setProcessedNumOfRecords(processedNumOfRecords);
+        updateProcessed(dataProcessing, operation, processedNumOfRecords);
+
         bulkOperationRepository.findById(operation.getId()).ifPresent(op -> operation.setCommittedNumOfErrors(op.getCommittedNumOfErrors()));
+      } catch (S3ClientException e) {
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:385]: It makes to rework this also guiding by https://github.com/folio-org/mod-bulk-operations/pull/316/files#r1855540575 and https://github.com/folio-org/mod-bulk-operations/pull/316/files#r1855542832.

File: src/main/java/org/folio/bulkops/service/DataExportJobUpdateService.java [hunk 7]
```diff
@@ -124,14 +127,20 @@ private void downloadOriginFileAndUpdateBulkOperation(BulkOperation operation, J
       }
       operation.setEndTime(LocalDateTime.ofInstant(jobUpdate.getEndTime().toInstant(), UTC_ZONE));
 
+    } catch (S3ClientException e) {
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/service/DataExportJobUpdateService.java:130]: It makes to rework this also guiding by https://github.com/folio-org/mod-bulk-operations/pull/316/files#r1855540575 and https://github.com/folio-org/mod-bulk-operations/pull/316/files#r1855542832.

File: src/test/java/org/folio/bulkops/service/BulkOperationServiceTest.java [hunk 8]
```diff
@@ -601,7 +601,7 @@ void  shouldPopulateErrorToBulkOperationIfS3IssuesForConfirmChangesForInstanceMa
         .thenReturn(BulkOperationDataProcessing.builder()
```
Comment (khandramai) [inline src/test/java/org/folio/bulkops/service/BulkOperationServiceTest.java:601]: Given the number of places in the code where changes have been made, it is essential to add or enhance existing tests to emulate the occurrence of both S3 exceptions and errors arising from other causes.

**Review by alekGbuz — APPROVED** [review]

**Pull request merged**